usenix87/Utilities/Date/Fdate/fdate.1
.TH FDATE 1
.if n .ds lq '
.if n .ds rq '
.if t .ds lq `
.if t .ds rq '
.MU
.SH NAME
fdate \- format date to standard output
.SH SYNOPSIS
.B
fdate [-n]
.BR [-F c ]
args
.SH DESCRIPTION
.I Fdate
writes its arguments to standard output
just like
.IR echo (1).
With the option
.B \-n
the newline normally supplied at the end is supressed.
.PP
However, unlike echo,
.I fdate
recognizes certain patterns in its arguments, and replaces them
by various parts of the current date and time.
.PP
Each conversion specification is introduced by
the character
.BR % .
(If the option
.BR \-F c
is used, then the character
.I c
replaces the
.BR % .)
Any other characters are simply copied to the output.
Following the
.BR % ,
there may be
.TP
\-
an optional minus sign
.RB "\*(lq" "\-" "\*(rq"
which specifies
.I "left adjustment"
of the converted value
in the
indicated field;
.TP
\-
an optional digit string specifying a
.I "field width;"
if the converted value has fewer characters
than the field width
it will be blank-padded on the left (or right,
if the left-adjustment indicator has been
given) to make up the field width;
if the field width begins with a zero,
zero-padding will be done instead of blank-padding;
.TP
\-
an optional period
.RB \&\*(lq . \&\*(rq
which serves to
separate the field width from the
next digit string;
.TP
\-
an optional digit string
specifying a
.I maximum
.I width
which specifies
the maximum number of characters
to be printed from this conversion;
.TP
\-
an optional plus sign
.RB \*(lq + \*(rq
which serves to introduce the next digit string;
.TP
\-
an optional digit string
specifying an increment to be applied to the current time.
The increment is in units indicated by the conversion character.
Only
.RB \*(lq y \*(rq,
.RB \*(lq m \*(rq,
.RB \*(lq d \*(rq,
.RB \*(lq H \*(rq,
.RB \*(lq M \*(rq,
and
.RB \*(lq S \*(rq
may have increments, others are ignored.
Increments are applied in increasing order of signifigance, and
all are applied before any output.
.TP
\-
an optional second plus sign
.RB \*(lq + \*(rq
indicating that this conversion is to produce no output.
.TP
\-
a character which indicates the value to be inserted.
.PP
The conversion characters
and the values that replace them are
.TP
.B d
The day of the month (an integer from 1 to 31).
.TP
.B m
The month of the year (an integer from 1 to 12).
.TP
.B y
The current year (relative to the beginning of the century).
.TP
.B D
The day, month and year in the format: dd/mm/yy. If zero fill
was specified, then each of the three fields is guaranteed to
be exactly two digits wide (with a leading zero if necessary).
.TP
.B Y
The day, month, and year in the format: yy/mm/dd. Zero fill operates
in the same way as for
.BR %D .
.TP
.B H
The hour of the day.
.TP
.B M
The minute of the hour.
.TP
.B S
The second of the minute.
.TP
.B X
An approximation to the number of milliseconds past the second of
the minute.
.TP
.B T
The current time in the format: hh:mm:ss. Zero fill operates in the
same way as for
.BR %D .
.TP
.B R
The hour of the day, using a 12 hour clock (an integer between
1 and 12), in place of the 24 hour clock used by
.BR %H .
.TP
.B r
The current time in the format: hh:mm:ss XM, where
.I X
is either
.B A
or
.BR P .
This gives a twelve hour clock, in contrast to the 24 hour clock
obtained through
.BR %T .
Zero fill works in the same manner as for
.BR %D .
.TP
.B w
The day of the week (an integer between 0 and 6).
.TP
.B A
The name of the day of the week.
.TP
.B a
The first three characters of the name of the day of the week
(ie: equivalent to
.BR %.3A ).
.TP
.B L
The name of the current month.
.TP
.B l
The first three characters of the name of the current month.
.TP
.B j
The julian day (day of the year), an integer between 1 and 366.
.TP
.B z
The name of the local timezone (as produced by
.BR timezone (3)).
.TP
.B C
A 24 character string containing the current time as formatted by
.BR ctime (3),
but without the trailing newline or null.
.TP
.B p
Either
.B am
or
.BR pm ,
as appropriate.
.TP
.B P
Either
.B AM
or
.BR PM ,
as appropriate
.RB ( %p 
in capitals).
.TP
.B h
One of the four strings: \*(lqst\*(rq, \*(lqnd\*(rq, \*(lqrd\*(rq, or \*(lqth\*(rq, whichever
is appropriate for the current day of the month.
.TP
.B %
A \*(lq%\*(rq; the field width, justification, fill, and maximum characters
indications are ignored if specified.
.TP
.B n
A newline (widths etc. are ignored).
.TP
.B t
A tab (widths are ignored here too).
.PP
A small field width will never cause truncation of a value, however
a small maximum width value will. A maximum width parameter larger
than the width of the value is ignored, but a large field width
causes padding (on the left or right, with spaces or zeroes).
.PP
.B Examples
.br
To print the date in the same format that
.BR date (1)
uses:
.IP
fdate %.19C %z 19%02y
.PP
To make a C program segment, containing a character string
with the current date (for compilation date stamping):
.RS
.nf
.sp
fdate 'char version [\^] = {%n%t"prog version nn, %0Y %0T"%n};'
.fi
.RE
.PP
To set a shell variable to the current hour of the day:
.IP
var=`fdate %H`
.PP
To run some commands in an hour from now:
.IP
at `fdate %02+1H%02M`
.br
commands
.br
^D
.SH "SEE ALSO"
ctime(3),
date(1),
echo(1),
sh(1)
.SH BUGS
There is no %Z for zodiac signs.
.br
There is no way to go back in time.
(Though if the day, month, etc. are not wanted,
going forward 23 hours and 55 minutes is close enough
to going back 5 minutes).
.br
Going forward a month from January 31st
is not what everybody expects.
.br
Going forward around about the time that daylight saving turns
off and on could also be considered strange.
.br
On Bell System V, the \fB%X\fP conversion always returns 0.


usenix87/Utilities/Date/Fdate/fdate.c
static char sccsid[] = "@(#)fdate.c	4.3 (Melbourne) 82/01/25";

/*
 *	fdate - formatted date output
 */

#include        <stdio.h>
#include	<sys/types.h>
#include	<time.h>

#ifdef SysV
struct timeb {
	time_t	time;
	int	millitm;
	int	timezone;
};
#else
#include	<sys/timeb.h>
#endif

#define	MONTH	itoa(tim->tm_mon+1,p)
#define	DAY	itoa(tim->tm_mday,p)
#define	YEAR	itoa(tim->tm_year,p)
#define	HOUR	itoa(tim->tm_hour,p)
#define	MINUTE	itoa(tim->tm_min,p)
#define	SECOND	itoa(tim->tm_sec,p)
#define	JULIAN	itoa(tim->tm_yday+1,p)
#define	WEEKDAY	itoa(tim->tm_wday,p)
#define	MODHOUR	itoa(h,p)
#define	MILISEC	itoa(timeb.millitm,p)

#define	FILL(num) {if (fill == '0' && tim->tm_/**/num < 10) *p++ = '0';}

char	*month[] = {
	"January",
	"February",
	"March",
	"April",
	"May",
	"June",
	"July",
	"August",
	"September",
	"October",
	"November",
	"December",
};

int	monthdays[12] = {
	31,
	28,
	31,
	30,
	31,
	30,
	31,
	31,
	30,
	31,
	30,
	31
};

char	*days[] = {
	"Sunday",
	"Monday",
	"Tuesday",
	"Wednesday",
	"Thursday",
	"Friday",
	"Saturday",
};

char	*suffix[] = {
	"th",
	"st",
	"nd",
	"rd",
	"th",
	"th",
	"th",
	"th",
	"th",
	"th",
};

char	*itoa();
#ifndef SysV
char	*timezone();
#endif
struct	tm	*localtime();
char	*asctime();
int	nonl;

struct	tm	incr	= { 0 };



main(argc, argv)
	int	argc;
	char	**argv;
{
	register char	*aptr, c;
	struct	timeb	timeb;
	register struct	tm  *tim;
	register int i, h, hflag;
	char *p;
	int	ndig, nchar, ljust;
	char	fill;
	char	lead;
	int	hadincr;
	int	ac;
	char	**av;
	char	 buf[60];
	extern	char	_sobuf[];

	setbuf(stdout, _sobuf);

#ifdef SysV
	tzset();
#endif

	ftime(&timeb);
	tim = localtime(&timeb.time);

	lead = '%';

	while (argc > 1 && argv[1][0] == '-') {
		switch (argv[1][1]) {

		case 'n':
			if (argv[1][2] == '\0') {
				nonl++;
				argv++;
				argc--;
				continue;
			}
			break;

		case 'F':
			if (argv[1][2] != '\0' && argv[1][3] == '\0') {
				lead = argv[1][2];
				argv++;
				argc--;
				continue;
			}
			break;
		}
		break;
	}

	ac = argc;
	av = argv;

	hadincr = 0;
	while (--ac > 0) {
		aptr = *++av;

		while (c = *aptr++) {
			if (c != lead)
				continue;

			i = 0;

			if (*aptr == '-')
				aptr++;
			while (*aptr >= '0' && *aptr <= '9')
				aptr++;
			if (*aptr == '.')
				while (*++aptr >= '0' && *aptr <= '9')
					;
			if (*aptr == '+') {
				while (*++aptr >= '0' && *aptr <= '9')
					i *= 10, i += *aptr - '0';
				if (*aptr == '+')
					aptr++;
			}
			hadincr++;

			switch (*aptr++) {

			case 'd':	incr.tm_mday = i;	break;
			case 'm':	incr.tm_mon = i;	break;
			case 'y':	incr.tm_year = i;	break;

			case 'H':	incr.tm_hour = i;	break;
			case 'M':	incr.tm_min = i;	break;
			case 'S':	incr.tm_sec = i;	break;

			case '\0':	aptr--;		/* fall through */
			default:	hadincr--;		break;
			}
		}
	}

	if (hadincr) {
		tim->tm_sec += incr.tm_sec;
		while (tim->tm_sec >= 60) {
			tim->tm_min++;
			tim->tm_sec -= 60;
		}

		tim->tm_min += incr.tm_min;
		while (tim->tm_min >= 60) {
			tim->tm_hour++;
			tim->tm_min -= 60;
		}

		tim->tm_hour += incr.tm_hour;
		while (tim->tm_hour >= 24) {
			tim->tm_mday++;
			tim->tm_hour -= 24;
		}

		setyear(tim->tm_year);
		tim->tm_mday += incr.tm_mday;
		while (tim->tm_mday > monthdays[tim->tm_mon]) {
			tim->tm_mday -= monthdays[tim->tm_mon];
			if (++tim->tm_mon >= 12) {
				tim->tm_mon -= 12;
				setyear(++tim->tm_year);
			}
		}

		tim->tm_mon += incr.tm_mon;
		while (tim->tm_mon >= 12) {
			tim->tm_year++;
			tim->tm_mon -= 12;
		}

		tim->tm_year += incr.tm_year;
		setyear(tim->tm_year);

		if (tim->tm_mday > monthdays[tim->tm_mon])
			tim->tm_mday = monthdays[tim->tm_mon];

		timeb.time = 0;
		tim->tm_year += 1900;

		for (i = 1970; i < tim->tm_year; i++)
			timeb.time += dysize(i);

		while (tim->tm_mon)
			timeb.time += monthdays[--tim->tm_mon];

		timeb.time += tim->tm_mday - 1;
		timeb.time *= 24;
		timeb.time += tim->tm_hour;
		timeb.time *= 60;
		timeb.time += tim->tm_min;
		timeb.time *= 60;
		timeb.time += tim->tm_sec;

		timeb.time += (long)timeb.timezone*60;
		if (tim->tm_isdst)
			timeb.time -= 60*60;
		tim = localtime(&timeb.time);
	}

	while (--argc > 0) {
		aptr = *++argv;

		while (c = *aptr++) {
			for (p = buf; p < &buf[60]; )
				*p++ = '\0';

			ndig = nchar = ljust = 0;
			fill = ' ';
			p = buf;
			if (c == lead) {
				if (*aptr == '-') {
					aptr++;
					ljust++;
				}
				if (*aptr == '0') {
					fill = '0';
					aptr++;
				}
				i = 0;
				while (*aptr >= '0' && *aptr <= '9') {
					i *= 10;
					i += *aptr++ - '0';
				}
				ndig = i;
				if (*aptr == '.') {
					i = 0;
					while (*++aptr >= '0' && *aptr <= '9')
						i *= 10, i += *aptr - '0';
					nchar = i;
				}
				if (*aptr == '+') {
					while (*++aptr >= '0' && *aptr <= '9')
						;
					if (*aptr == '+') {
						aptr++;
						if (*aptr)
							aptr++;
						continue;
					}
				}
				switch (c = *aptr++) {
				case '%':
					putchar('%');
					continue;
				case 'n':
					putchar('\n');
					continue;
				case 't':
					putchar('\t');
					continue;
				case 'p':
					if (tim->tm_hour >= 12)
						p = "pm";
					else
						p = "am";
					strcpy(buf, p);
					break;
				case 'P':
					if (tim->tm_hour >= 12)
						p = "PM";
					else
						p = "AM";
					strcpy(buf, p);
					break;
				case 'h':
					i = tim->tm_mday;
					if (i > 10 && i < 19)
						i = 5;
					i %= 10;
					strcpy(buf, suffix[i]);
					break;
				case 'm' :
					p = MONTH;
					break;
				case 'd':
					p = DAY;
					break;
				case 'y' :
					p = YEAR;
					break;
				case 'Y':
					FILL(year);
					p = YEAR;
					*p++ = '/';
					FILL(mon+1);
					p = MONTH;
					*p++ = '/';
					FILL(mday);
					p = DAY;
					break;
				case 'D':
					FILL(mday);
					p = DAY;
					*p++ = '/';
					FILL(mon+1);
					p = MONTH;
					*p++ = '/';
					FILL(year);
					p = YEAR;
					break;
				case 'H':
					p = HOUR;
					break;
				case 'M':
					p = MINUTE;
					break;
				case 'S':
					p = SECOND;
					break;
				case 'T':
					FILL(hour);
					p = HOUR;
					*p++ = ':';
					FILL(min);
					p = MINUTE;
					*p++ = ':';
					FILL(sec);
					p = SECOND;
					break;
				case 'j':
					p = JULIAN;
					break;
				case 'w':
					p = WEEKDAY;
					break;
				case 'R':
					h = tim->tm_hour;
					if ((h %= 12) == 0)
						h = 12;
					p = MODHOUR;
					break;
				case 'r':
					if ((h = tim->tm_hour) >= 12)
						hflag++;
					if ((h %= 12) == 0)
						h = 12;
					if (fill == '0' && h < 10)
						*p++ = '0';
					p = MODHOUR;
					*p++ = ':';
					FILL(min);
					p = MINUTE;
					*p++ = ':';
					FILL(sec);
					p = SECOND;
					*p++ = ' ';
					if (hflag)
						*p++ = 'P';
					else 
						*p++ = 'A';
					*p++ = 'M';
					break;
				case 'l':
					if (nchar == 0 || nchar > 3)
						nchar = 3;
				case 'L':
					p = month[tim->tm_mon];
					strcpy(buf, p);
					break;
				case 'a':
					if (nchar == 0 || nchar > 3)
						nchar = 3;
				case 'A':
					p = days[tim->tm_wday];
					strcpy(buf, p);
					break;
				case 'X':
					p = MILISEC;
					break;
				case 'C':
					if (nchar == 0 || nchar > 24)
						nchar = 24;
					p = asctime(tim);
					strcpy(buf, p);
					break;
				case 'z':
#ifdef SysV
					p = tzname[tim->tm_isdst];
#else
					p = timezone(timeb.timezone,tim->tm_isdst);
#endif
					strcpy(buf, p);
					break;
				default:
					if (c == lead) {
						putchar(c);
						break;
					}
					if (c == '\0' || c == ' ')
						fprintf(stderr, "fdate: missing format character in \"%s\"\n", *argv);
					else
						fprintf(stderr, "fdate: bad format character - %c\n", c);
					close(1);	/* trash stdout! */
					exit(2);

				}	/* endsw */
			} else {
				putchar(c);
				continue;
			}

			h = strlen(buf);
			if (nchar && nchar < h)
				h = nchar;
			i = ndig - h;
			if (!ljust)
				while (i-- > 0)
					putchar(fill);
			p = buf;
			while (h-- > 0)
				putchar(*p++);
			while (i-- > 0)
				putchar(fill);

		}	/* endwh */

		if (argc > 1)
			putchar(' ');
	}	/* endwh */

	if (!nonl)
		putchar('\n');
	exit(0);
}

char *
itoa(i, ptr)
	register  int	i;
	register  char	*ptr;
{
	char	buf[8];
	register	char *p = buf;
	do {
		*p++ = i % 10 + '0';
		i /= 10;
	} while(i);

	while (p > buf)
		*ptr++ = *--p;
	*ptr = '\0';
	return (ptr);
}

setyear(yr)
	register yr;
{
	if (yr > 2100) {
		fprintf(stderr, "fdate: time got too big\n");
		exit(1);
	}

	if (dysize(yr) == 366)
		monthdays[1] = 29;
	else
		monthdays[1] = 28;
}

#ifdef SysV

ftime(p)
	struct timeb *p;
{
	(void) time (&p->time);
	p->millitm = 0;
	p->timezone = timezone / 60;
}

int
dysize(yr)
	register yr;
{
	if (yr % 4)
		return (365);
	if (yr % 100)
		return (366);
	if (yr % 400)
		return (365);
	return (366);
}

#endif
